### PR TITLE
Supporting memory efficient dropout in flash attention

### PR DESCRIPTION
**Pull request by tongxin**
The main difference from Tri Dao's cuda implementation is the way we handle philox rng state.  We cannot easily control the per thread philox offset increment since Triton makes thread abstraction opaque. That means we cannot reproduce the dropout masks exactly the same way as the cuda version. 
The second difference is about the nuances of where to apply scaling, masking, casting and everything. It's an experience not guided by principles but rather trial and error.

### CODE REVIEW

**Review by tongxin — COMMENTED** [review]
Looks good.

File: src/flag_attn/dropout.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    BLOCK: tl.constexpr = 1024
+    offs_base += tl.arange(0, BLOCK)
+    for start_n in range(0, N, BLOCK):
+        offs = start_n + offs_base
+        rng_offs = offset + offs
+        pmask = tl.rand(seed, rng_offs, n_rounds=6) > dropout_p
+        row_mask = start_n + tl.arange(0, BLOCK) < N
+        tl.store(dropout_mask + offs, pmask, mask=row_mask)
+
+def dropout_mask(x, B, H, M, N, dropout_p, seed, offset):
```
Comment (iclementine) [inline src/flag_attn/dropout.py:30]: Is this function only used for testing for now?

Comment (iclementine) [reply src/flag_attn/dropout.py:30]: Also, if x is only used to get the device, may be replacing it with a explicit `device` argument is better.

Comment (tongxin) [reply src/flag_attn/dropout.py:30]: That's a good suggestion.

Comment (tongxin) [reply src/flag_attn/dropout.py]: Updated.

**Review by iclementine — APPROVED** [review]
LGTM

**Review by iclementine — APPROVED** [review]

**Pull request merged**